tls/s2n_tls13_secrets.c

```diff
@@ -0,0 +1,501 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+#include "tls/s2n_tls13_secrets.h"
+
+#include "tls/s2n_connection.h"
+#include "tls/s2n_key_log.h"
+#include "tls/s2n_tls13_handshake.h"
+#include "utils/s2n_bitmap.h"
+
+#define S2N_MAX_HASHLEN SHA384_DIGEST_LENGTH
+
+#define CONN_HMAC_ALG(conn) ((conn)->secure.cipher_suite->prf_alg)
+#define CONN_SECRETS(conn)  ((conn)->secrets.tls13)
+#define CONN_HASHES(conn)   ((conn)->handshake.hashes)
+
+#define CONN_SECRET(conn, secret) ( \
+    (struct s2n_blob) { .data = CONN_SECRETS(conn).secret, .size = s2n_get_hash_len(CONN_HMAC_ALG(conn))} )
+#define CONN_HASH(conn, hash) ( \
+    (struct s2n_blob) { .data = CONN_HASHES(conn)->hash, .size = s2n_get_hash_len(CONN_HMAC_ALG(conn))} )
+#define CONN_FINISHED(conn, mode) ( \
+    (struct s2n_blob) { .data = (conn)->handshake.mode##_finished, .size = s2n_get_hash_len(CONN_HMAC_ALG(conn))})
+
+/**
+ *= https://tools.ietf.org/rfc/rfc8446#section-7.1
+ *# If a given secret is not available, then the 0-value consisting of a
+ *# string of Hash.length bytes set to zeros is used.
+ */
+static uint8_t zero_value_bytes[S2N_MAX_HASHLEN] = { 0 };
+#define ZERO_VALUE(hmac_alg) ( \
+    (const struct s2n_blob) { .data = zero_value_bytes, .size = s2n_get_hash_len(hmac_alg)})
+
+/**
+ * When an operation doesn't need an actual transcript hash,
+ * it uses an empty transcript hash as an input instead.
+ *
+ *= https://tools.ietf.org/rfc/rfc8446#section-7.1
+ *# Note that in some cases a zero-
+ *# length Context (indicated by "") is passed to HKDF-Expand-Label
+ */
+#define EMPTY_CONTEXT(hmac_alg) ( \
+    (const struct s2n_blob) { .data = s2n_get_empty_context(hmac_alg), .size = s2n_get_hash_len(hmac_alg)})
+
+static uint8_t s2n_get_hash_len(s2n_hmac_algorithm hmac_alg)
+{
+    uint8_t hash_size = 0;
+    if (s2n_hmac_digest_size(hmac_alg, &hash_size) != S2N_SUCCESS) {
+        return 0;
+    }
+    return hash_size;
+}
+
+static uint8_t *s2n_get_empty_context(s2n_hmac_algorithm hmac_alg)
+{
+    static uint8_t sha256_empty_digest[S2N_MAX_HASHLEN] = { 0 };
```

Comment:
They're not actually the same size. I just allocate the maximum hash size for both because paranoia.

But more importantly, the digests themselves aren't just empty arrays. The digest of nothing is actually something! That's why we call the init method.